### ISSUE TITLE
Thoughts on "camera units"?

### ISSUE THREAD

**Issue by timholy** [ajkeller34/Unitful.jl#53]
This is really just a discussion topic, it's not something I think belongs in this package. If you don't find it interesting, just close it.

I'm mulling over the proper framework for expressing values recorded from scientific cameras. Such cameras typically report values as, e.g., `UInt16`. The existing `FixedPointNumbers` framework allows one to normalize these numbers to a 0-to-1 scale, where 1 means "saturated" (which is something quite important to keep track of---you want to know when your sensor has been maxed out). For example, a 12-bit camera might return values in the range `0x0000:0x0fff`, and representing those as `N4f12` numbers (where `0x0fff` corresponds to 1) is a good choice because it places what would otherwise be arbitrary integers onto a common scale possessing genuine physical meaning.

However, scientific cameras often have another important feature: there's often a known conversion from "camera units" to "number of photons." This, too, is a very meaningful quantity, since the number of photons gives an estimate of the uncertainty in the intensity measurement (`sqrt(N)` due to Poisson statistics).

So here we have a raw measurement, expressed in essentially arbitrary units, that might meaningfully be reinterpreted in two distinct ways. This sort of feels like physical units (we can convert microns to inches to light-picoseconds), but it also seems a bit different. It kind of seems to blur the lines between dimensionless numbers and dimensionful numbers. One of the reasons I'm thinking about Unitful is the very nice framework you've developed for custom units, but I think there are conceptual as well as practical issues here. Any particular insights on the right choice of framework for representing such numbers? I'm CCing @rsrock and @Cody-G to see if they have any interest in the discussion topic.

**Comment by ajkeller34** [comment]
Great discussion topic. First, let me point out that there is support for dimensionless, unitful numbers. This is useful in e.g. distinguishing between degrees and radians. At first glance, you could imagine defining a `nphotons` unit and a `camera` unit for each camera model (both dimensionless, with camera quantities backed by an `N4f12` pure numeric type), and then have methods dispatch accordingly. 

Then again, I guess it's probably number of photons per integration time, so that already hints that these things are not dimensionless. I guess the camera is only collecting photons that fall within some bandwidth too. Not using cameras at that level of detail, does it make sense to go between photons per time and power, i.e. do the counted photons have a well-defined energy? Maybe these are really power units, and the camera units are just scaled so that 1 is some threshold power. I'm just spitballing.

This also reminds me of frequencies: sometimes in physics papers you can only tell if they mean rad/s or Hz if they use an obvious symbol, like `f` or `omega`. If they are quoting a coupling rate, often they will just write `2π * whatever Hz` to wink at you and indicate that this is an angular frequency. A while back I added a `Hz2π` unit for this situation, which has a built-in 2π conversion factor. It's not clear that this is the best approach---why not just do `rad*Hz`---but I kind of don't like having radians hang around.

Unitful really doesn't care if your units and dimensions are physical, of course. You can think of units as tags to dispatch on numbers differently, and dimensions as classes of fungible tags.

**Comment by timholy** [comment]
> First, let me point out that there is support for dimensionless, unitful numbers.

That may be the best way to go. Thanks for pointing this out, I had forgotten.

>  does it make sense to go between photons per time and power, i.e. do the counted photons have a well-defined energy?

Good thoughts. It depends on the application, but typically no. Modern cameras are basically photon-counting devices: each absorbed photon results in the promotion of a single electron, which then gets converted by the camera gain (an arbitrary and sometimes even tunable parameter) into the digital number the camera spits out. There's definitely a limited spectrum of light that can cause electron promotion, but it can be more than a two-fold range of wavelength/energy and within that spectrum all photons are indistinguishable in terms of the output.

I'll explore the dimensionless units. Thanks for your help! (I'll close this just so it doesn't hang around, but others should of course feel free to chime in.)

**Comment by timholy** [comment]
Hmm, there's an interesting issue with making them dimensionless. I've pushed a tiny package, [CameraUnits](https://github.com/timholy/CameraUnits.jl), so one can look at code if desired.

```julia
julia> using CameraUnits, Unitful

julia> x = 1cu"pco55"
1 cu

julia> uconvert(γe, x)
0.46 γe

julia> using FixedPointNumbers

julia> convert(N0f16, x)
0.46N0f16      # not what I want
```
When converting to normalized units, I'd want the answer
```julia
julia> reinterpret(N0f16, 0x0001)
2.0e-5N0f16
```
But the presence of an automatic conversion implies that, even if I wrote a specialized `convert` method, I could get myself into conflicts by the infamous two-step conversions.

Is the better approach to add a new dimension? It's rather weird because I'd like to be able to convert to `N0f16`, which is dimensionless.

**Comment by timholy** [comment]
I should also have made it clear that the number of photoelectrons per digital number is not a universal constant; it depends on the gain. So one 16-bit camera might use 0.46, another 1.24, etc. So there's no way to convert `γe` to a normalized unit like `N0f16`.

**Comment by ajkeller34** [comment]
After some further thought, I think that because ultimately you're capturing a number of photons per integration time, this is a quantity with dimensions 1/[time]. 

It would be unphysical to ignore that the photons are being captured in some time interval. Even if the integration time were fixed by the camera it would still make sense to convert to a frequency (how many photons per second were there, during the time this frame was recorded). Also, presumably different cameras may have different integration times, so directly comparing "number of photons" for a given captured frame between cameras seems problematic. 

With that in mind, to get to the 0 to 1 scale, it seems to me like you normalize by dividing out the threshold number of photons/sec for a given camera, which therefore implies the zero to one scale should be dimensionless. You would have to define conversion constants: a threshold number of photons/sec for each camera. If I started with `0.5N0f16`, and I had defined e.g. `const pco55_cu = 0.46γe/[integration time]`, then I could just multiply as `(0.5N0f16)cu"pco55"` and get the number of photons/sec. There's no unit conversion happening, per se: you're just multiplying by a dimensionful constant.

Now suppose you want to get the integer out, given some photons/sec. Just divide by the constant for that camera. You'll wind up with a number, probably between 0 and 1 but maybe not if your photons / sec were more than the threshold for that camera. If it's in range you can use `N0f16` to convert from `Float64`, followed by `reinterpret` to get the integer code.

Finally, this is a good opportunity to mention that `mol` has dimensions [number] in SI. At least as implemented in Unitful right now, you go between dimensionless and dimensionful counts by dividing/multiplying the Avogadro constant, which has units `1/mol`. It seems like this is somehow a similar situation, where you have two quantities that seem unit-convertible but have different dimensions. Which reminds me, you might consider whether you'd also want [number] for your photon counting. It may be formally correct, but it would make conversion between photons/s and `Hz` needlessly annoying since `Hz` has dimensions 1/[time], not [number]/[time].

**Comment by timholy** [comment]
I agree with that perspective. But I think that both the rate and the total quantity (integral of rate) are independently important. For example, it's only the integral that matters for the purposes of estimating the noise in your measurements. If someone gives me the values reported by a camera, tells me the conversion factor to photoelectrons, but doesn't happen to know the sampling rate at which the images were collected, I can still do useful stuff with the images.

A second factor is that, while I enthusiastically endorse your thinking here of moving this towards invariant underlying physical processes, this isn't the only factor that comes in to play. For astronomical imaging, the measured quantity also depends on pixel area (photons/sec/μm^2), the magnification of your optical system (this can be combined with the previous factor), the numeric aperture of the optical system, and the quantum efficiency of the detector. In biological fluorescence imaging, it moreover depends on the brightness of your illumination (e.g., laser power), concentration of fluorophores, and their fluorescence lifetime. All this combines to say that 'directly comparing "number of photons" for a given captured frame between cameras seems problematic' wouldn't be solved by just encoding the rate. Indeed, in biological imaging, if I have one experiment for which I'm happy with a framerate of 5Hz, but another in which I want a framerate of 100Hz, I will likely turn up the laser intensity by a factor of 20 for the latter experiment, because what I want to do is match the collected number of photons. In other words, at least in my world that's the closest thing to a useful invariant.

**Comment by timholy** [comment]
I'm beginning to think this is a more general problem than I posed at the beginning, though still one that's surely a separate package rather than something to add to Unitful. Consider a setup in which you have a microphone connected to an amplifier connected to an A/D card. The A/D will ultimately report `Int16`s or something, and overflow is something you'd like to be aware of. (In other words, the representation as an `Int16` matters.) But the A/D clearly has a conversion between these digital numbers and volts...and together the microphone + amplifier have a conversion between volts and Pa. The conventional approach is to just have separate variables storing these conversion factors, and you pass them around as extra function arguments whenever you need them. But in Julia we have the opportunity---if we decide to take it---to encode these conversion factors in the type system so that they are carried around with the numbers themselves. For example, just to make it interesting let's say you have an [18-bit ADC](http://www.linear.com/product/LTC2387-18) that reports values as `Int32`; I'm deliberately choosing something where the significant bits don't match the width of a built-in numeric type, so as to show off how FixedPointNumbers can help. We might handle this with

```julia
julia> immutable ADCNumber{T<:Number,conversions} <: Number
           val::T
       end

# A numeric type for 18-bit signed, ADC range of +/- 10V, and a microphone/amplifier gain of 5.0mV/Pa
julia> T = ADCNumber{Q13f18, (10V/reinterpret(typemax(Q13f18)), 5.0mV/Pa)}
ADCNumber{FixedPointNumbers.Fixed{Int32,18},(4.656612875245797e-9 V,5.0 Pa^-1 mV)}

# The following methods would need to be defined
x = reinterpret(T, Int32(-345))   # -345 is the number returned by the ADC
real(x) # would return reinterpret(Q13f18, Int32(-345)) == -0.001316Q13f18
V(x)    # would return -1.6065314419598e-6 V
Pa(x)   # would return -0.00032130628839196 Pa
```

This is basically the same problem as the one I posed at the top, except that it had only one meaningful conversion number (to photoelectrons).

In some ways it's an abuse of the type system to encode these conversion factors: it would be horrifying if people had thousands of different types of numbers, all differing only by different gain factors. (The amount of compilation boggles the mind, and just storing all that compiled code in memory would bring the entire system crashing to the earth.) But if you are collecting large arrays of all the same type of number (including the same conversion factors), this seems more justifiable.

Reactions? Worth also CCing @bjarthur here as the main author of [NIDAQ](https://github.com/JaneliaSciComp/NIDAQ.jl) and someone who's participated in the design of FixedPointNumbers.

**Comment by ajkeller34** [comment]
Re: your first post, yes, other dimensions would come into a full accounting that I ignored. I'll just go with counting photons for further discussion then.

Perhaps Unitful can address your most recent post already, if you put the conversions in method definitions instead of defining types that store the conversion factors. See the following (note: you'll want to update Unitful master since I just addressed a macro problem and also got builds to pass again, and possibly be on julia master since `where` syntax seems to change with some regularity). I've changed some numbers from your CameraUnits.jl package for sake of example:

```
# with up-to-date julia 0.6 and Unitful...
using Unitful, FixedPointNumbers

@dimension 📷 "📷" CameraTag #... that's a camera emoji if it doesn't render right
@refunit reftag "*" RefCameraTag 📷 false
@unit γe       "γe"  PhotoElectron      1          false
@unit pco55_cu "(pco55)" PCO55 1reftag false
@unit pco42_cu "(pco42)" PCO42 1reftag false

# define methods for uconvert to go between dimensions
Unitful.uconvert(::typeof(γe), x::Quantity{<:FixedPoint,typeof(📷),typeof(pco42_cu)}) =x*0.42γe/pco42_cu
Unitful.uconvert(::typeof(γe), x::Quantity{<:FixedPoint,typeof(📷),typeof(pco55_cu)}) =x*0.55γe/pco55_cu

# some other methods you could imagine defining...
Unitful.uconvert(::typeof(γe), x::Quantity{UInt16,typeof(📷),typeof(pco42_cu)}) =reinterpret(N0f16, x/pco42_cu)*0.42γe
Unitful.uconvert(::typeof(γe), x::Quantity{<:Number,typeof(📷),typeof(pco42_cu)}) = error("ambiguous. does 65535 mean 0xffff? Or is it 65535*the full range?")

# If you're going backward from photoelectrons to camera units, I'm going to assume you want something with the precision / range of the camera.
Unitful.uconvert(::typeof(pco42_cu), x::Quantity{<:Number,typeof(NoDims),typeof(γe)}) = N0f16(x/0.42γe)*pco42_cu
Unitful.uconvert(::typeof(pco55_cu), x::Quantity{<:Number,typeof(NoDims),typeof(γe)}) =N0f16(x/0.55γe)*pco55_cu
```

Now let's try this out:

```
julia> 0xffff*pco42_cu   # tagged a UInt16
0xffff (pco42)

julia> ans |> γe
0.42 γe

julia> reinterpret(N0f16, 0xabcd)*pco55_cu  # tagged a FixedPointNumber
0.67111N0f16 (pco55)

julia> ans |> γe
0.3691088731212329 γe

julia> photons = ans
0.3691088731212329 γe

julia> photons |> pco55_cu
0.67111N0f16 (pco55)

julia> photons |> pco42_cu
0.87883N0f16 (pco42)

julia> 4*pco42_cu |> γe
ERROR: ambiguous. does 65535 mean 0xffff? Or is it 65535*the full range?
Stacktrace:
 [1] |>(::Quantity{Int64, Dimensions:{📷}, Units:{(pco42)}}, ::Unitful.Units{(Unitful.Unit{:PhotoElectron,Unitful.Dimensions{()}}(0, 1//1),),Unitful.Dimensions{()}}) at ./operators.jl:872
```

Now, one could just as easily write conversions to whatever other dimensions you want, e.g. if it made sense to identify the camera-tagged numbers with other physical quantities.

There's certainly some magic going on in deciding what you mean by a given numeric type, which could be viewed as convenient or scary. I also take liberties with the meaning of unit conversion. It's now become a tag swapping operation in the case of camera tags. Also, it's non-standard to allow unit conversions between dimensions, although we're maintaining full control over when that happens (see [this related issue](https://github.com/ajkeller34/Unitful.jl/issues/37)).

Hopefully this is stimulating if nothing else.
